### ISSUE TITLE
`local let` warns about unused variable `let`.

### ISSUE THREAD

**Issue by cjwelborn** [koalaman/shellcheck#450]
I was messing around trying to declare a `local` variable using `let` at the same time. It seems there are two ways to do this, and `shellcheck` gives a false positive for one of them. I don't know what the 'best-practice' is, but `local let` seems to work as it should.

With a line like this:
```bash
local let x=1
```

`shellcheck` should be checking `x`, not `let`.

My test script:

```bash
# Setup some globals, to see if the 'local' works.
let x=2600
let y=1337
let z=808

function testfunc() {
	# This variable will not be local.
	let local x=0
	# Shellcheck complains about this line !!
	# shellcheck disable=SC2034
	local let y=1
	# Workaround, to ease the complaint.
	local z=$(( 2 ))
}
testfunc

function check_global {
	printf "Outer %s: %4s - " "$3" "$1"
	if [[ "$1" == "$2" ]]; then
		echo "Correct, using \`$4\`"
	else
		echo "Global was modified using \`$4\`!"
	fi
}

check_global "$x" 2600 "x" "let local"
check_global "$y" 1337 "y" "local let"
check_global "$z" 808 "z" "\$(())"
```

The output:
```
Outer x:    0 - Global was modified using `let local`!
Outer y: 1337 - Correct, using `local let`
Outer z:  808 - Correct, using `$(())`
```

**Comment by deryni** [comment]
`local let a=1` does in fact define a `let` variable though and doesn't evaluate arguments arithmetically.

Compare the output of

    f() { declare -p let a; local a=5+1; declare -p let a; local; }

to the output of

    f() { declare -p let a; local let a=5+1; declare -p let a; local; }

and the output of

    f() { declare -p let a; let a=5+1; declare -p let a; local; }

and

    f() { local a; declare -p let a; let a=5+1; declare -p let a; local; }

**Comment by deryni** [comment]
Your test script shows that `let local` doesn't set the variable as local (the `x` test) and shows that `local let` and `local` both do set a variable locally. But it asserts nothing about the value of `let` or the operation of `let` (the builtin).

**Comment by cjwelborn** [comment]
I was just investigating this. `let` is a valid variable name, and you're right. I should've tested whether the `let` operation was doing what it was supposed to.

```bash
function locallet {
	local let x=1+1
	echo " local let x=1+1: $x"
	local y=$((1+1))
	echo "local x=\$((1+1)): $y"
}
locallet
```

Output:
```
 local let x=1+1: 1+1
local x=$((1+1)): 2
```

Sorry for wasting your time. I don't know how many people have tried this, but it's probably not many. A google search for "local let" doesn't turn up very much. Otherwise I would request a lint for this along the lines of "local let does not do what you think it does".I would be the only one to ever see it though :wink:.

Now I know what not to do.

**Comment by deryni** [comment]
Yeah, I don't know whether that's common enough to warrant being called out specifically but that's not my decision. =)

**Comment by koalaman** [comment]
A github search finds several examples of this, like [here](https://github.com/kkonopko/kriscience/blob/e494ce58d280bd10061a7e0dfa336d908e686c07/xmldsig-c14n/test-c14n.sh#L54), [here](https://github.com/ielia/db-patch-automata/blob/4a01d1a50af31e47b55a21aca51d9c776ee149ea/env_var_function#L8) and [here](https://github.com/songgang/SongPipeline/blob/f1d9a5975e2b64195dfed961540a0172a3d9a335/eduardo_ILD/feature_selection/spireg.sh#L41).

It's still pretty rare and the error message you get is correct, so I just added a sentence to the [wiki](https://github.com/koalaman/shellcheck/wiki/SC2034) for now.

**Comment by deryni** [comment]
Huh, interesting. I wonder what the hit rate on the things shellcheck *does* catch would be when run over all (or some large sub-set) of the scripts available on github would be. That would be neat to find out.

**Comment by koalaman** [comment]
I do this regularly on an informal corpus of scripts scraped from IRC (so mostly scripts/snippets people are asking for help with), and in that case I consider 10+ cases per 1000 scripts frequent, and 1-3 cases per 1000 scripts infrequent but useful. For this instance, I found 0 cases in 5000 scripts.

A github corpus would be very interesting, since it would say more about the type of errors people aren't aware of.

**Comment by deryni** [comment]
Yeah, github and StackOverflow/CodeReview/etc. would be interesting as corpora of data for the varying levels of awareness implied of the participants. Getting the data isn't the simplest thing in the world though. =)

**Comment by cjwelborn** [comment]
I'm sure it would be huge. I'm always looking at other people's code. Some are well known projects, and some are not so well known. I use either `Sublime Text` or `Atom`, which both have plugins for `shellcheck`. I don't see that many errors, but the warnings are everywhere.

Out of 89 open source projects (not my own, found in my `~/clones` directory) I found 1566 unique `.sh` files with warnings. Out of those 1566 files there were 16808 total warnings.

The command I used: `find -type f -name "*.sh" -exec "shellcheck" '{}' ';' | ./shellcheckall.sh`

The [./shellcheckall.sh](https://welbornprod.com/view?file=/static/files/scripts/shellcheckall.sh) script was something I threw together to get a quick count, although that `find` command is not so quick on this many directories :wink: .

The [raw output file](https://welbornprod.com/view?file=/static/files/text/shellcheck-clone-errors.txt) is about 3Mb.

My check includes projects like Git, Rust, Cargo (Rust's package manager), Python, HHVM, coreutils, XChat (not maintained anymore, HexChat is the best fork), and even one small warning in Bats (the Bash Automated Test System) which was 'declare and assign separately'.

Most warnings are 'Double quote to prevent expansion', 'Use `$( )` instead of `` ` ` ``', and 'Declare and assign separately to avoid masking return values'.

**Comment by deryni** [comment]
Using `-exec shellcheck {} \+` to run shellcheck on multiple files (in a single directory) at once will probably help the speed of your `find` run a bit. Also those download links seem to not work for me.

**Comment by cjwelborn** [comment]
Thanks for the tip. The links above are for my 'file viewer', which requires javascript. I'm not sure why they wouldn't work, but here are the straight download links: 

[shellcheckall.sh](https://welbornprod.com/dl/static/files/scripts/shellcheckall.sh)

[raw output](https://welbornprod.com/dl/static/files/text/shellcheck-clone-errors.txt)

**Comment by aw-altiscale** [comment]
FWIW, we've built shellcheck support in what will be a new Apache project to handle some of the precommitment work that ASF committers have to do.  (All the excitement is happening over on https://issues.apache.org/jira/browse/HADOOP-12111 if you are interested).  

As a result, we've been running shellcheck on lots of the ASF repos.  The amount of output generated by shellcheck is pretty staggering, mainly due to the large amounts of copypasta.

**Comment by koalaman** [comment]
@aw-altiscale Awesome! Be sure to let me know of any pain points you discover in this kind of integration.
